### DISCUSSION TITLE
Virtual folder as the home dir

### DISCUSSION THREAD

**Discussion by sergejostir** [drakkan/sftpgo#387] [Ideas]
Hello.

Is there a chance that setting virtual folder as a home dir would be possible? For example, if you have users that can only access shared (virtual) folders. Currently, they are logged in their "home" folder and then have to make additional click to enter the virtual folder that they can actually use.

**Comment by drakkan** [comment]
So basically you want to set a start directory instead of `/`. Can you please confirm that I understand correctly?

I think this it is not possibile for WebDAV while it could work for FTP/SFTP. I think pkg/sftp does not currently support this feature. I'll try to investigate and to add this support

**Reply by sergejostir** [reply]
I was thinking more like that you can set a virtual folder as "home_dir". Currently "home_dir" has to physically exist on the filesystem, so if you are only using virtual folders then you have to keep an empty dir somewhere to set "home_dir" to it.

**Reply by drakkan** [reply]
Then I don't understand. Let's assume we're using the local filesystem for simplicity (it is the same for other backends).

You have a virtual folder pointing to the local path `/tmp/home`.
You can simply set the home dir for a given user to `/tmp/home` and don't set the virtual folder at all.

My initial understanding was that you have a user with the home dir as `/tmp/user` and a virtual folder mounted on `/vdir` and you want to use `/vdir` as starting directory and to build relative paths (basically the response for `SSH_FXP_REALPATH` requests) but still allow the user to use an absolute path like `/` to use his home dir.

**Reply by drakkan** [reply]
Here is my first attempt to add start directory support in pkg/sftp

https://github.com/pkg/sftp/pull/437

**Reply by sergejostir** [reply]
> Then I don't understand. Let's assume we're using the local filesystem for simplicity (it is the same for other backends).
> 
> You have a virtual folder pointing to the local path `/tmp/home`.
> You can simply set the home dir for a given user to `/tmp/home` and don't set the virtual folder at all.
> 
> My initial understanding was that you have a user with the home dir as `/tmp/user` and a virtual folder mounted on `/vdir` and you want to use `/vdir` as starting directory and to build relative paths (basically the response for `SSH_FXP_REALPATH` requests) but still allow the user to use an absolute path like `/` to use his home dir.

Ah, that makes sense. I thought I would have to use virtual folders for my use case, but since I'm using external authentication I can actually use users as "virtual users" because of https://github.com/drakkan/sftpgo/issues/125.

I'm going to map multiple physical persons with different logins to the same sftpgo username. My testing showed that all user information is stored in connection/session, so they can all have different permissions even though they are using the same username internally.

**Reply by drakkan** [reply]
Yes, this usage should work, the user info are never reloaded after the login. You have to disable quota (it will be wrong) and maybe you have also to disable password caching if you are using a recent development snapshot.

While that is not what you are asking for, I think adding a starting directory might be a good addition for FTP/SFTP, I will try to implement it before 2.1

**Reply by sergejostir** [reply]
> Yes, this usage should work, the user info are never reloaded after the login. You have to disable quota (it will be wrong) and maybe you have also to disable password caching if you are using a recent development snapshot.

Thanks for confirmation. I will be periodically updating quota anyway using https://github.com/drakkan/sftpgo/issues/130 which is applied to the current connections.
What do you mean with password caching?

**Reply by drakkan** [reply]
> > Yes, this usage should work, the user info are never reloaded after the login. You have to disable quota (it will be wrong) and maybe you have also to disable password caching if you are using a recent development snapshot.
> 
> Thanks for confirmation. I will be periodically updating quota anyway using #130 which is applied to the current connections.
> What do you mean with password caching?

I mean this [setting](https://github.com/drakkan/sftpgo/blob/main/sftpgo.json#L174)

**Reply by sergejostir** [reply]
> I mean this [setting](https://github.com/drakkan/sftpgo/blob/main/sftpgo.json#L174)

But is it used when using external auth?

**Reply by drakkan** [reply]
> > I mean this [setting](https://github.com/drakkan/sftpgo/blob/main/sftpgo.json#L174)
> 
> But is it used when using external auth?

yes
